docs/repo/crates/discv4.md

```diff
@@ -0,0 +1,316 @@
+# Discv4
+
+The `discv4` crate plays an important role in Reth, enabling discovery of other peers across the network. It is recommended to know how [Kademlia distributed hash tables](https://en.wikipedia.org/wiki/Kademlia) and [Ethereum's node discovery protocol](https://github.com/ethereum/devp2p/blob/master/discv4.md#pong-packet-0x02) work before reading through this chapter. While all concepts will be explained through the following sections, reading through the links above will make understanding this chapter much easier! With that note out of the way, lets jump into `disc4`.
+
+## Starting the Node Discovery Protocol
+As mentioned in the network and stages chapters, when the node is first started up, the `node::Command::execute()` function is called, which initializes the node and starts to run the Reth pipeline. Throughout the initialization of the node, there are many processes that are are started. One of the processes that is initialized is the p2p network which starts the node discovery protocol amongst other tasks.  
+
+[File: bin/reth/src/node/mod.rs]()
+```rust ignore
+  pub async fn execute(&self) -> eyre::Result<()> {
+    //--snip--
+    let network = config
+        .network_config(db.clone(), chain_id, genesis_hash, self.network.disable_discovery)
+        .start_network()
+        .await?;
+
+    info!(peer_id = ?network.peer_id(), local_addr = %network.local_addr(), "Started p2p networking");
+
+    //--snip--
+    }
+```
+
+During this process, a new `NetworkManager` is created through the `NetworkManager::new()` function, which starts the discovery protocol through a handful of newly spawned tasks. Lets take a look at how this actually works under the hood. 
+
+[File: ]()
+```rust ignore
+impl<C> NetworkManager<C>
+where
+    C: BlockProvider,
+{
+    //--snip--
+
+    pub async fn new(config: NetworkConfig<C>) -> Result<Self, NetworkError> {
+        let NetworkConfig {
+            //--snip--
+            secret_key,
+            mut discovery_v4_config,
+            discovery_addr,
+            boot_nodes,
+            //--snip--
+            ..
+        } = config;
+
+        //--snip--
+
+        discovery_v4_config = discovery_v4_config.map(|mut disc_config| {
+            // merge configured boot nodes
+            disc_config.bootstrap_nodes.extend(boot_nodes.clone());
+            disc_config.add_eip868_pair("eth", status.forkid);
+            disc_config
+        });
+
+        let discovery = Discovery::new(discovery_addr, secret_key, discovery_v4_config).await?;
+
+    //--snip--
+    }
+}
+```
+
+First, the `NetworkConfig` is deconstructed and the `disc_config` is updated to merge configured [bootstrap nodes]() and add the `forkid` to adhere to [EIP 868](https://eips.ethereum.org/EIPS/eip-868). This updated configuration variable is then passed into the `Discovery::new()` function. Note that `Discovery` is a catch all for all discovery services, which include discv4, DNS discovery and others in the future.
```

Comment:
as I was reading through I realized we don't actually see how the boot nodes get used, might be cool to include (not necessary to do so, I was just curious)